src/Gaffer/Spreadsheet.cpp

```diff
@@ -0,0 +1,557 @@
+	}
+	return result;
+}
+
+//////////////////////////////////////////////////////////////////////////
+// RowPlug
+//////////////////////////////////////////////////////////////////////////
+
+GAFFER_PLUG_DEFINE_TYPE( Spreadsheet::RowPlug );
+
+Spreadsheet::RowPlug::RowPlug( const std::string &name, Plug::Direction direction, unsigned flags )
+	:	ValuePlug( name, direction, flags )
+{
+	addChild( new StringPlug( "name", direction ) );
+	addChild( new BoolPlug( "enabled",direction, true ) );
+	addChild( new ValuePlug( "cells", direction ) );
+}
+
+StringPlug *Spreadsheet::RowPlug::namePlug()
+{
+	return getChild<StringPlug>( 0 );
+}
+
+const StringPlug *Spreadsheet::RowPlug::namePlug() const
+{
+	return getChild<StringPlug>( 0 );
+}
+
+BoolPlug *Spreadsheet::RowPlug::enabledPlug()
+{
+	return getChild<BoolPlug>( 1 );
+}
+
+const BoolPlug *Spreadsheet::RowPlug::enabledPlug() const
+{
+	return getChild<BoolPlug>( 1 );
+}
+
+ValuePlug *Spreadsheet::RowPlug::cellsPlug()
+{
+	return getChild<ValuePlug>( 2 );
+}
+
+const ValuePlug *Spreadsheet::RowPlug::cellsPlug() const
+{
+	return getChild<ValuePlug>( 2 );
+}
+
+Gaffer::PlugPtr Spreadsheet::RowPlug::createCounterpart( const std::string &name, Direction direction ) const
+{
+	RowPlugPtr result = new RowPlug( name, direction );
+	for( const auto &cell : CellPlug::Range( *cellsPlug() ) )
+	{
+		result->cellsPlug()->addChild( cell->createCounterpart( cell->getName(), direction ) );
+	}
+	return result;
+}
+
+//////////////////////////////////////////////////////////////////////////
+// CellPlug
+//////////////////////////////////////////////////////////////////////////
+
+GAFFER_PLUG_DEFINE_TYPE( Spreadsheet::CellPlug );
+
+Spreadsheet::CellPlug::CellPlug( const std::string &name, const Gaffer::Plug *value, Plug::Direction direction )
+	:	ValuePlug( name, direction )
+{
+	addChild( new BoolPlug( "enabled", direction, true ) );
+	addChild( value->createCounterpart( "value", direction ) );
+	valuePlug()->setFlags( Plug::Dynamic, false );
+}
+
+BoolPlug *Spreadsheet::CellPlug::enabledPlug()
+{
+	return getChild<BoolPlug>( 0 );
+}
+
+const BoolPlug *Spreadsheet::CellPlug::enabledPlug() const
+{
+	return getChild<BoolPlug>( 0 );
+}
+
+Gaffer::PlugPtr Spreadsheet::CellPlug::createCounterpart( const std::string &name, Direction direction ) const
+{
+	return new CellPlug( name, valuePlug(), direction );
+}
+
+//////////////////////////////////////////////////////////////////////////
+// Spreadsheet
+//////////////////////////////////////////////////////////////////////////
+
+size_t Spreadsheet::g_firstPlugIndex = 0;
+GAFFER_GRAPHCOMPONENT_DEFINE_TYPE( Spreadsheet );
+
+Spreadsheet::Spreadsheet( const std::string &name )
+	:	ComputeNode( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new BoolPlug( "enabled", Plug::In, true ) );
+	addChild( new StringPlug( "selector" ) );
+	addChild( new RowsPlug( "rows" ) );
```

Comment:
I don't think this makes sense in the constructor. It would make subsequent scripting a bit painful :

```
s = Spreadsheet()
for r in rowsIWouldLikeToAdd() :
   if itsTheFirst() :
       row = s["rows"][1]
   else :
       row = s["rows"].addRow()
```

I do think it makes sense in the UI though, hence why `SpreadsheetUI.__createSpreadsheet()` creates a row automatically. Maybe I should do the same for the item in the NodeMenu? I was wondering about removing the menu item though, at least while the only way of creating columns is in the plug menu, so it's always easier to create a spreadsheet from there. What do you think?

Reply:
Sounds good to me. Especially re: having one way to make a spreadsheet. Also fits the expression pattern, in the sense of plug-centric use.